rc/src/lib.rs

```diff
@@ -37,17 +46,63 @@ impl<T> MyRc<T> {
         MyRc {
             ptr: Box::into_raw(Box::new(t)),
             refcount: Box::into_raw(Box::new(1)),
+            weakcount: Box::into_raw(Box::new(0)),
         }
     }
     
     pub fn consume(self) -> Result<T,MyRc<T>> {
         match unsafe{*self.refcount} {
             1 => {
+                    match unsafe{*self.weakcount} {
+                        0 => {
+                            unsafe{Box::from_raw(self.refcount);}
+                            unsafe{Box::from_raw(self.weakcount);}
+                        },
+                        _ => unsafe{std::ptr::write(self.refcount, *self.refcount - 1);},

```

Comment:
You actually don't need to use `write` here. It would be equivalent to assignment.